core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java

```diff
@@ -66,6 +104,129 @@ public void checkRead(Object key) {
+      boolean canRollback = availabilityMode == AvailabilityMode.AVAILABLE &&
+            !getLastStableTopology().getActualMembers().contains(globalTransaction.getAddress());
+      if (trace) {
+         log.tracef("Can rollback transaction? %s", canRollback);
+      }
+      return canRollback;
+   }
+
+   @Override
+   public void onTopologyUpdate(CacheTopology cacheTopology) {
+      boolean isStable = isTopologyStable(cacheTopology);
+      if (trace) {
+         log.tracef("On stable topology update. has pending transactions? %b. Is stable? %b. topology=%s",
+                    !partialTransactions.isEmpty(), isStable, cacheTopology);
+      }
+      if (isStable) {
+         for (TransactionInfo transactionInfo : partialTransactions.values()) {
+            completeTransaction(transactionInfo, cacheTopology);
+         }
+      }
+   }
+
+   private void completeTransaction(final TransactionInfo transactionInfo, CacheTopology cacheTopology) {
+      final Collection<Address> affectedNodes = transactionInfo.getCommitNodes(cacheTopology);
+      NotifyingNotifiableFuture<Map<Address, Response>> notifyingFuture = new NotifyingFutureImpl<>();
+      notifyingFuture.attachListener(new FutureListener<Map<Address, Response>>() {
+         @Override
+         public void futureDone(Future<Map<Address, Response>> future) {
+            final GlobalTransaction globalTransaction = transactionInfo.getGlobalTransaction();
+            if (trace) {
+               log.tracef("Future done for transaction %s", globalTransaction);
+            }
+            try {
+               future.get();
+            } catch (Throwable e) {
+               if (trace) {
+                  log.tracef(e, "Exception for transaction %s", globalTransaction);
+               }
+               return;
+            }
+            if (trace) {
+               log.tracef("Performing cleanup for transaction %s", globalTransaction);
+            }
+            lockManager.unlock(transactionInfo.getLockedKeys(), globalTransaction);
+            partialTransactions.remove(globalTransaction);
+            TxCompletionNotificationCommand command = commandsFactory.buildTxCompletionNotificationCommand(null, globalTransaction);
+            rpcManager.invokeRemotely(affectedNodes, command, rpcManager.getDefaultRpcOptions(false));
+         }
+      });
+      rpcManager.invokeRemotelyInFuture(notifyingFuture, affectedNodes, transactionInfo.buildCommand(commandsFactory, isVersioned), rpcManager.getDefaultRpcOptions(true));
+   }
+
+   private boolean isTopologyStable(CacheTopology cacheTopology) {
+      CacheTopology stableTopology = localTopologyManager.getStableCacheTopology(cacheName);
+      if (trace) {
+         log.tracef("Check if topology %s is stable. Last stable topology is %s", cacheTopology, stableTopology);
+      }
+      return stableTopology != null && cacheTopology.getActualMembers().containsAll(stableTopology.getActualMembers());
```

Comment:
We could change the merge procedure to update the stable topology as well, if you think it would help. We could also rename `onTopologyUpdate()` to `onStableTopologyUpdate()` and call it directly from `LocalTopologyManagerImpl.handleStableTopologyUpdate()`...